builder/codebuild/codebuild.go

```diff
@@ -0,0 +1,353 @@
+		err = <-done
+		if err == nil {
+			err = ctx.Err()
+		}
+	case err = <-done:
+	}
+
+	if err == context.Canceled || err == context.DeadlineExceeded {
+		return &builder.BuildCanceledError{
+			Err:    err,
+			Reason: ctx.Err(),
+		}
+	}
+
+	return err
+}
+
+func (b *Builder) tailBuild(ctx context.Context, w io.Writer, buildID *string) error {
+	fmt.Fprintf(w, "conveyor: Waiting for CodeBuild build %s to start\n", *buildID)
+
+	build, err := b.waitUntilLogsAvailable(ctx, buildID)
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprintf(w, "conveyor: CodeBuild build %s started, streaming logs...\n", *buildID)
+
+	r, err := b.logReader(*build.Logs.GroupName, *build.Logs.StreamName)
+	if err != nil {
+		return fmt.Errorf("opening cloudwatch logs stream for %s: %v", *build.Id, err)
+	}
+
+	copyDone := make(chan error)
+	go func() {
+		_, err := io.Copy(w, r)
+		copyDone <- err
+	}()
+
+	ctxTimeout, cancel := context.WithTimeout(context.Background(), BuildCompleteTimeout)
+	defer cancel()
+
+	build, _ = b.waitUntilBuildComplete(ctxTimeout, build.Id)
+
+	// Close the cloudwatch log stream reader.
+	r.Close()
+
+	// Wait for the io.Copy to complete flushing data to w.
+	if err := <-copyDone; err != nil {
+		return fmt.Errorf("copying logs: %v", err)
+	}
+
+	fmt.Fprintf(w, "conveyor: CodeBuild build %s completed\n", *buildID)
+
+	for _, status := range failedStatuses {
+		if *build.BuildStatus == status {
+			return fmt.Errorf("CodeBuild build %s failed", *buildID)
+		}
+	}
+
+	return nil
+}
+
+func (b *Builder) waitUntilLogsAvailable(ctx context.Context, buildID *string) (*codebuild.Build, error) {
+	input := &codebuild.BatchGetBuildsInput{Ids: []*string{buildID}}
+	resp, err := b.waitUntil(ctx, input, func(resp *codebuild.BatchGetBuildsOutput) bool {
+		build := resp.Builds[0]
+		return build.Logs != nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return resp.Builds[0], nil
+}
+
+func (b *Builder) waitUntilBuildComplete(ctx context.Context, buildID *string) (*codebuild.Build, error) {
+	input := &codebuild.BatchGetBuildsInput{Ids: []*string{buildID}}
+	resp, err := b.waitUntil(ctx, input, func(resp *codebuild.BatchGetBuildsOutput) bool {
+		build := resp.Builds[0]
+		return *build.BuildComplete
+	})
+	if err != nil {
+		return nil, err
+	}
+	return resp.Builds[0], nil
+}
+
+func (b *Builder) waitUntil(ctx context.Context, input *codebuild.BatchGetBuildsInput, fn func(*codebuild.BatchGetBuildsOutput) bool) (*codebuild.BatchGetBuildsOutput, error) {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+		}
+
+		resp, err := b.codebuild.BatchGetBuilds(input)
+		if err != nil {
+			return resp, err
+		}
+
+		if fn(resp) {
+			return resp, nil
+		}
+
+		time.Sleep(BatchGetBuildsWaitTime)
+	}
+}
+
+func (b *Builder) startBuild(ctx context.Context, opts BuildOptions) (*codebuild.StartBuildOutput, error) {
+	buf := new(bytes.Buffer)
+	if err := b.BuildSpec.Execute(buf, opts); err != nil {
```

Comment:
This is actually part of the stdlib: [text/template#Template.Execute](https://golang.org/pkg/text/template/#Template.Execute)